src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -4621,6 +4621,12 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
                     std::string var_name = std::string(v->m_name);
                     throw SemanticError("Assignment to loop variable `" + std::string(to_lower(var_name)) +"` is not allowed", target->base.loc);
                 }
+                if (sym->type == ASR::symbolType::Variable) {
+                    ASR::Variable_t *v = ASR::down_cast<ASR::Variable_t>(sym);
+                    if (v->m_intent == ASR::intentType::In) {
+                        throw SemanticError("Assignment to function parameter `" + std::string(v->m_name) + "` is not allowed", target->base.loc);
+                    }
+                }
```

Comment:
Well there is one case like,

```python
def f(x: ClassA) -> i32:
    x.Member1 = 2
    return x.Member1 * x.Member1
```

Should the above error or work?

Reply:
I think/guess it should work. I think it is one of the ways to pass variables/values by reference in python. Objects are passed by reference in python. In the above example, we pass the object `x` and inside the function, a data member of `x` is updated. After the call to the function `f`, the changes to data members of `x` would be visible to the caller function.

For example, consider the following:
```python
$ cat examples/expr2.py   
from lpython import i32, dataclass

@dataclass
class N:
    val: i32

def f1(n: i32):
    n = 2
    return

def f2(n: N):
    n.val = 2
    return

def main0():
    a: i32
    b: N = N(0)
    a = 4
    b.val = 4

    f1(a)
    f2(b)

    print(a)
    print(b.val)

main0()
$ python examples/expr2.py 
4
2
$ lpython examples/expr2.py
2
2
```

Reply:
```
def f(x: ClassA) -> i32:
    x.Member1 = 2
    return x.Member1 * x.Member1
```
We have to decide what the default passing is, whether: `In`, `InOut`, `Out`. I would start with `In`, so the above is not allowed. To allow it, do:
```
def f(x: InOut[ClassA]) -> i32:
    x.Member1 = 2
    return x.Member1 * x.Member1
```
Then it is allowed.

Reply:
See https://github.com/lcompilers/lpython/issues/1807.

Reply:
In the following example:
```python
def f(x: In[ClassA]) -> i32:
    x.Member1 = 2 # Assignment to Member1 which is a local variable of `x`
    x = ClassA(Member1=5, ...) # Assignment to x which is an intent In variable
    return x.Member1 * x.Member1
```
Do we print errors for both the above assignments or just for the second assignment (whose target variable has intent as `In`)?

Reply:
If in case we print error for the first assignment `x.Member1 = 2 # Assignment to Member1 which is a local variable of x`, do we need to print error in `CPython` as well? (I think/guess by default cpython allows assignment to class member variables).

Reply:
> Do we print errors for both the above assignments or just for the second assignment (whose target variable has intent as In)?

I would print error for both. "In" means that "both the pointer and what it points to is constant". If you want to assign to any of it, you have to mark it "InOut".

Regarding CPython, I would not worry about it. The contract is: if LPython compiles it, it will run in CPython also. So if LPython enforces no modification, then CPython doesn't need to enforce it. In CPython we simply ignore type annotations.